src/systemd.c

```diff
@@ -0,0 +1,272 @@
+					     "StartTransientUnit");
+	if (ret < 0) {
+		cgroup_err("failed to create the systemd msg: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_message_append(msg, "ss", scope_name, modes[opts->mode]);
+	if (ret < 0) {
+		cgroup_err("failed to append the scope name: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_message_open_container(msg, 'a', "(sv)");
+	if (ret < 0) {
+		cgroup_err("failed to open container: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_message_append(msg, "(sv)", "Description", "s",
+				    "scope created by libcgroup");
+	if (ret < 0) {
+		cgroup_err("failed to append the description: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_message_append(msg, "(sv)", "PIDs", "au", 1, child_pid);
+	if (ret < 0) {
+		cgroup_err("failed to append the PID: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_message_append(msg, "(sv)", "Slice", "s", slice_name);
+	if (ret < 0) {
+		cgroup_err("failed to append the slice: %d\n", errno);
+		goto out;
+	}
+
+	if (opts->delegated == 1) {
+		ret = sd_bus_message_append(msg, "(sv)", "Delegate", "b", 1);
+		if (ret < 0) {
+			cgroup_err("failed to append delegate: %d\n", errno);
+			goto out;
+		}
+	}
+
+	ret = sd_bus_message_close_container(msg);
+	if (ret < 0) {
+		cgroup_err("failed to close the container: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_message_append(msg, "a(sa(sv))", 0);
+	if (ret < 0) {
+		cgroup_err("failed to append aux structure: %d\n", errno);
+		goto out;
+	}
+
+	ret = sd_bus_call(bus, msg, 0, &error, &reply);
+	if (ret < 0) {
+		cgroup_err("sd_bus_call() failed: %d\n",
+			   sd_bus_message_get_errno(msg));
+		cgroup_err("error message: %s\n", error.message);
+		goto out;
+	}
+
+	/* Receive the job_path from systemd */
+	ret = sd_bus_message_read(reply, "o", &job_path);
+	if (ret < 0) {
+		cgroup_err("failed to read reply: %d\n", errno);
+		goto out;
+	}
+
+	cgroup_dbg("job_path = %s\n", job_path);
+
+	ret = clock_gettime(CLOCK_MONOTONIC, &start);
+	if (ret < 0) {
+		cgroup_err("Failed to get time: %d\n", errno);
+		ret = ECGFAIL;
+		goto out;
+	}
+
+	/* The callback will null out the job_path pointer on completion */
+	while(job_path) {
+		ret = sd_bus_process(bus, NULL);
+		if (ret < 0) {
+			cgroup_err("failed to process the sd bus: %d\n", errno);
+			goto out;
+		}
+
+		if (ret == 0) {
+			/*
+			 * Per the sd_bus_wait() man page, call this function after sd_bus_process
+			 * returns zero. The wait time (usec) was somewhat arbitrarily chosen
+			 */
+			ret = sd_bus_wait(bus, 10);
+			if (ret < 0) {
+				cgroup_err("failed to wait for sd bus: %d\n", errno);
+				goto out;
+			}
+		}
+
+		ret = clock_gettime(CLOCK_MONOTONIC, &now);
+		if (ret < 0) {
+			cgroup_err("Failed to get time: %d\n", errno);
+			ret = ECGFAIL;
+			goto out;
+		}
+
+		if (elapsed_time(&start, &now) > ONE_SEC) {
+			cgroup_err("The create scope command timed out\n");
+			ret = ECGFAIL;
+			goto out;
+		}
+	}
+
+	ret = 0;
+
+out:
```

Comment:
Should we kill the fork of `libcgroup_systemd_idle_thread` created by us on failure `kill (child_pid, SIGTERM);` ?

Reply:
Yes, but we need to be careful not to kill the customer's process.  Here's what I added:
```
        if (ret && opts->pid < 0)
                kill(child_pid, SIGTERM);
```

Reply:
yeah, that was the intention :)